driver-core/src/main/com/mongodb/operation/AsyncChangeStreamBatchCursor.java

```diff
@@ -0,0 +1,193 @@
+        wrapped.setBatchSize(batchSize);
+    }
+
+    @Override
+    public int getBatchSize() {
+        return wrapped.getBatchSize();
+    }
+
+    @Override
+    public boolean isClosed() {
+        return wrapped.isClosed();
+    }
+
+    @Override
+    public void close() {
+        wrapped.close();
+    }
+
+    private interface AsyncBlock {
+        void apply(AsyncQueryBatchCursor<RawBsonDocument> cursor, SingleResultCallback<List<RawBsonDocument>> callback);
+    }
+
+    private void resumeAbleOperation(final AsyncBlock asyncBlock, final SingleResultCallback<List<RawBsonDocument>> callback) {
+        boolean msgInFlight = inFlight.getAndSet(true);
+        if (msgInFlight) {
+            callback.onResult(null,
+                    new MongoChangeStreamException("Another asynchronous operation on this Changes Stream is already in progress"));
+            return;
+        }
+
+        asyncBlock.apply(wrapped, new SingleResultCallback<List<RawBsonDocument>>() {
+            @Override
+            public void onResult(final List<RawBsonDocument> result, final Throwable t) {
+                inFlight.set(false);
+                if (t == null || !changeStreamOperation.isResumable()) {
+                    callback.onResult(result, t);
+                } else if (t instanceof MongoServerException && !(t instanceof MongoNotPrimaryException
+                        || t instanceof MongoQueryException)) {
+                    callback.onResult(null, t);
+                } else {
+                    killCursorAndRetryOperation(asyncBlock, callback);
+                }
+            }
+        });
+    }
+
+    private void killCursorAndRetryOperation(final AsyncBlock asyncBlock, final SingleResultCallback<List<RawBsonDocument>> callback) {
+        ServerCursor serverCursor = wrapped.getServerCursor();
+        if (serverCursor == null) {
+            retryOperation(asyncBlock, callback);
+        } else {
+            MongoNamespace namespace = changeStreamOperation.getNamespace();
+            BsonDocument command = new BsonDocument("killCursors", new BsonString(namespace.getCollectionName()))
+                    .append("cursors", new BsonArray(singletonList(new BsonInt64(serverCursor.getId()))));
+            executeWrappedCommandProtocolAsync(binding, namespace.getDatabaseName(), command, BSON_DOCUMENT_CODEC,
+                    new SingleResultCallback<BsonDocument>() {
+                        @Override
+                        public void onResult(final BsonDocument result, final Throwable t) {
+                            // Ignore any exceptions killing old cursors
+                            retryOperation(asyncBlock, callback);
+                        }
+                    });
+        }
+    }
+
+    private void retryOperation(final AsyncBlock asyncBlock, final SingleResultCallback<List<RawBsonDocument>> callback) {
+        new ChangeStreamOperation<T>(changeStreamOperation, resumeToken)
```

Comment:
When retrying we create a new top level ChangeStreamOperation (which isn't resumable) and execute that. That allows us to modify the underlying BatchCursor.